# Ejercicios Clase: Álgebra en Ciencia de Datos
### Estudiante: Edgard Iglesias Rubio

## Unidad 3
### Aproximar valores propios

1.Lea sobre el método de potencias para aproximar el valor porpio mayor de una matríz. Realice los dos ejercicios que se encuentran al final de la guía:

#### Exercise 1: 
Let A be the matrix from the Inverse Power Method example: 


$$
A =
\begin{bmatrix}
9 & -1 & -3\\
0 & 6 & 0\\
-6 & 3 & 6
\end{bmatrix}
$$

(a) Use the Power Method to approximate the largest eigenvalue λ1. Verify that the exact value of λ1 is 12.

In [1]:
#imports
import numpy as np
from scipy.linalg import lu_factor, lu_solve

In [2]:
A = np.array([[9, -1, -3],[0, 6, 0],[-6, 3, 6]])
X = np.array([[1],[0],[0]])
i=0 #iteraciones
epsilon=0.0001
MAX_ITERACIONES=100

Y = A@X
diff = Y - np.linalg.norm(Y)*X

while (i<MAX_ITERACIONES and np.linalg.norm(diff)>epsilon):
    X=Y
    X = X/np.linalg.norm(X)
    
    Y = A@X
    diff = Y - np.linalg.norm(Y)*X
    
    i+=1
    
print("El mayor autovalor aproximadamente es: {lambd}".format(lambd=np.linalg.norm(Y)))

El mayor autovalor aproximadamente es: 12.000008583030988


(b) Apply the Inverse Power Method with a shift of μ=10. Explain why the results differ from those in the example.

In [3]:
A = np.array([[9, -1, -3],[0, 6, 0],[-6, 3, 6]])
X = np.array([[0],[1],[0]])
i=0 #iteraciones
epsilon=0.0001
MAX_ITERACIONES=100

diff=X
I=np.eye(3)
mu=10
A_shifted = A - mu*I
LU_factorization = lu_factor(A_shifted)

while (i<MAX_ITERACIONES and np.linalg.norm(diff)>epsilon):
    X_prev=X
    X=lu_solve(LU_factorization,X)
    X = X/np.linalg.norm(X)
    
    diff = X - X_prev
    
    i+=1
    
print("El autovalor aproximadamente es: {lambd}".format(lambd=np.linalg.norm(A@X)))

El autovalor aproximadamente es: 11.99997253020264


En este caso los autovalores de la matriz A son [12,3,6], al darle un parametro μ=10, convergió hacia el autovalor 12 ya que está mas cercano

(c) Apply the Inverse Power Method with a shift of μ=7.5 and the initial vector given below. Explain why the sequence of vectors approach the eigenvector corresponding to λ1

In [4]:
A = np.array([[9, -1, -3],[0, 6, 0],[-6, 3, 6]])
X = np.array([[1],[0],[0]])
i=0 #iteraciones
epsilon=0.0001
MAX_ITERACIONES=100

diff=X
I=np.eye(3)
mu=7.5
A_shifted = A - mu*I
LU_factorization = lu_factor(A_shifted)

while (i<MAX_ITERACIONES and np.linalg.norm(diff)>epsilon):
    X_prev=X
    X=lu_solve(LU_factorization,X)
    X = X/np.linalg.norm(X)
    
    diff = X - X_prev
    
    i+=1
    
print("El autovalor aproximadamente es: {lambd}".format(lambd=np.linalg.norm(A@X)))

El autovalor aproximadamente es: 10.816653826391969


Se aproximará a λ1 debido al vector X que se ha escogido que tenderá a hacer 0 los otros autovalores λ y buscará el primero más aproximado, en este caso λ1. Si se utilizara X=[0,1,0] bajo la misma configuracion del algoritmo se aproximaría a λ2

#### Exercise 2: 
Let B be the following matrix:


$$
A =
\begin{bmatrix}
-2 & -18 & 6\\
-11 & 3 & 11\\
-27 & 15 & 31
\end{bmatrix}
$$


(a) Apply the Power Method and Inverse Power Method with shifts to approximate all eigenvalues of the matrix B. (Note that one of the eigenvalues of this matrix is negative)

In [5]:
B = np.array([[-2, -18, 6],[-11, 3, 11],[-27, 15, 31]])
X = np.array([[1],[0],[0]])
i=0 #iteraciones
epsilon=0.0001
MAX_ITERACIONES=100

Y = B@X
diff = Y - np.linalg.norm(Y)*X

while (i<MAX_ITERACIONES and np.linalg.norm(diff)>epsilon):
    X=Y
    X = X/np.linalg.norm(X)
    
    Y = B@X
    diff = Y - np.linalg.norm(Y)*X
    
    i+=1
    
print("El mayor autovalor aproximadamente es: {lambd}".format(lambd=np.linalg.norm(Y)))

El mayor autovalor aproximadamente es: 35.9999767900754


In [6]:
B = np.array([[-2, -18, 6],[-11, 3, 11],[-27, 15, 31]])
X = np.array([[0],[1],[0]])
i=0 #iteraciones
epsilon=0.0001
MAX_ITERACIONES=100

diff=X
LU_factorization = lu_factor(B)

while (i<MAX_ITERACIONES and np.linalg.norm(diff)>epsilon):
    X_prev=X
    X=lu_solve(LU_factorization,X)
    X = X/np.linalg.norm(X)
    
    diff = X - X_prev
    
    i+=1
    
print("El autovalor aproximadamente es: {lambd}".format(lambd=np.linalg.norm(B@X)))

El autovalor aproximadamente es: 4.000549345743138


In [7]:
B = np.array([[-2, -18, 6],[-11, 3, 11],[-27, 15, 31]])
X = np.array([[0],[0],[1]])
i=0 #iteraciones
epsilon=0.0001
MAX_ITERACIONES=100

diff=X
I=np.eye(3)
mu=-32
B_shifted = B - mu*I
LU_factorization = lu_factor(B_shifted)

while (i<MAX_ITERACIONES and np.linalg.norm(diff)>epsilon):
    X_prev=X
    X=lu_solve(LU_factorization,X)
    X = X/np.linalg.norm(X)
    
    diff = X - X_prev
    
    i+=1
    
print("El autovalor aproximadamente es: {lambd}".format(lambd=np.linalg.norm(B@X)))

El autovalor aproximadamente es: 8.00360976533596


Los autovalores de B son aproximadamente: [32,4,-8]

(b) Check your results using the eig function in SciPy.

In [8]:
np.linalg.eig(B)

(array([36., -8.,  4.]),
 array([[ 1.02854624e-17, -8.16496581e-01,  7.07106781e-01],
        [-3.16227766e-01, -4.08248290e-01,  1.89360275e-16],
        [-9.48683298e-01, -4.08248290e-01,  7.07106781e-01]]))

2. Cácule los valores propios y los valores singulares para las siguientes matrices:

$$A =
\begin{bmatrix}
0 & 1 & 0 & 0\\
0 & 0 & 2 & 0\\
0 & 0 & 0 & 3\\
0 & 0 & 0 & 0
\end{bmatrix}
$$

$$
B = 
\begin{bmatrix}
0 & 1 & 0 & 0\\
0 & 0 & 2 & 0\\
0 & 0 & 0 & 3\\
0.001 & 0 & 0 & 0
\end{bmatrix}$$

In [9]:
A = np.array([[0, 1, 0, 0],[0, 0, 2, 0],[0, 0, 0, 3],[0, 0, 0, 0]])
#Valores singulares
_, Vsingulares, _ = np.linalg.svd(A)
#Valores propios
Vpropios=np.linalg.eigvals(A)

In [10]:
Vsingulares

array([3., 2., 1., 0.])

In [11]:
Vpropios

array([0., 0., 0., 0.])

A es una matriz singular cuyos valores propios son 0 y sus valores singulares: 3, 2, 1

In [12]:
B = np.array([[0, 1, 0, 0],[0, 0, 2, 0],[0, 0, 0, 3],[0.001, 0, 0, 0]])
#Valores singulares
_, Vsingulares, _ = np.linalg.svd(B)
#Valores propios
Vpropios=np.linalg.eigvals(B)

In [13]:
Vsingulares

array([3.e+00, 2.e+00, 1.e+00, 1.e-03])

In [14]:
Vpropios

array([-2.78315768e-01+0.j        , -7.97972799e-17+0.27831577j,
       -7.97972799e-17-0.27831577j,  2.78315768e-01+0.j        ])

B tiene los siguientes valores propios: -0.28,-7e-17 +0.28j, -7e-17 -0.28j, 0.28  y los siguientes valores singulares: 3, 2, 1